cpp/src/tests/sssp/sssp_test.cu

```diff
@@ -0,0 +1,419 @@
+
+    if (DoPreds) {
+      for (auto i = 0; i < v; ++i) {
+        for (auto e = vlist[i]; e < vlist[i + 1]; ++e) {
+          DistType weight = weights[e];
+          uint64_t key = (uint64_t)i << 32 | (uint64_t)elist[e];
+          if (min_edge_map.find(key) != min_edge_map.end()) {
+            min_edge_map[key] == std::min(weight, min_edge_map[key]);
+          } else {
+            min_edge_map[key] = weight;
+          }
+        }
+      }
+    }
+
+    ref_sssp(
+        vlist, elist, weights, param.src_, ref_distances, ref_predecessors);
+
+    for (auto i = 0; i < v; ++i) {
+      if (DoDist)
+        ASSERT_EQ(dist_vec[i], ref_distances[i])
+            << "vid: " << i << "ref dist " << ref_distances[i]
+            << " actual dist " << dist_vec[i];
+
+      if (DoPreds) {
+        if (pred_vec[i] != -1) {
+          uint64_t key = (uint64_t)pred_vec[i] << 32 | (uint64_t)i;
+          DistType min_edge_weight = min_edge_map.at(key);
+          ASSERT_EQ(ref_distances[pred_vec[i]] + min_edge_weight,
+                    ref_distances[i])
+              << "vid: " << i << "pred " << pred_vec[i] << " ref dist "
+              << ref_distances[i] << " observed " << ref_distances[pred_vec[i]]
+              << " + " << min_edge_weight << " = "
+              << ref_distances[pred_vec[i]] + min_edge_weight << "\n";
+        } else {
+          ASSERT_EQ(pred_vec[i], ref_predecessors[i])
+              << "vid: " << i << "ref pred " << ref_predecessors[i]
+              << " actual " << pred_vec[i];
+        }
+      }
+    }
+  }
+};
+
+std::vector<double> Tests_SSSP::SSSP_time;
+
+TEST_P(Tests_SSSP, CheckFP32_DIST_NO_PREDS) {
+  run_current_test<float, true, false>(GetParam());
+}
+TEST_P(Tests_SSSP, CheckFP32_NO_DIST_PREDS) {
+  run_current_test<float, false, true>(GetParam());
+}
+TEST_P(Tests_SSSP, CheckFP32_DIST_PREDS) {
+  run_current_test<float, true, true>(GetParam());
+}
+TEST_P(Tests_SSSP, CheckFP64_DIST_NO_PREDS) {
+  run_current_test<double, true, false>(GetParam());
+}
+TEST_P(Tests_SSSP, CheckFP64_NO_DIST_PREDS) {
+  run_current_test<double, false, true>(GetParam());
+}
+TEST_P(Tests_SSSP, CheckFP64_DIST_PREDS) {
+  run_current_test<double, true, true>(GetParam());
+}
+
+// --gtest_filter=*simple_test*
+
+INSTANTIATE_TEST_CASE_P(
+    simple_test,
+    Tests_SSSP,
+    ::testing::Values(
+        SSSP_Usecase(RMAT,
+                     "grmat --rmat_scale=10 --rmat_edgefactor=16 --device=0  "
+                     "--normalized --quiet",
+                     0),
+        SSSP_Usecase(RMAT,
+                     "grmat --rmat_scale=12 --rmat_edgefactor=8 --device=0  "
+                     "--normalized --quiet",
+                     10),
+        SSSP_Usecase(MTX, "test/datasets/dblp.mtx", 100),
+        SSSP_Usecase(MTX, "test/datasets/wiki2003.mtx", 100000),
+        SSSP_Usecase(MTX, "test/datasets/karate.mtx", 1)
+        //,SSSP_Usecase(MTX, "test/datasets/cit-Patents.mtx", 100)
```

Comment:
2 questions : 
- why cit-Patents.mtx is commented? We may want to delete or uncomment.
- In the previous sssp test suite, we were comparing against reference results stored in files on s3 for a couple more datasets in order to capture different scenario. Can you just double check that we are still testing both directed and undirected?

Reply:
I commented it since the validation runs (serial textbook cpu code) take a while, particularly for validating predecessors since it needs to create the min_edge_weight map. I can uncomment it if runtime is not a huge concern. I can optionally add other larger graphs in a separate case that does only performance benchmarks, but not validation. I ran other cases too and they all pass (with cuda-memcheck).

Reply:
Sounds good. The current test solution is fine if it captures all common cases.